File: src/nylon/parser/parsers/MacroParser.java
package nylon.parser.parsers;

import nylon.nylonobjects.NylonFunction;
import nylon.parser.CharIterator;
import nylon.parser.NylonParser;
import nylon.parser.ParseUtils;
import nylon.parser.Parser;

import java.util.ArrayList;

/**
 * Created by Aedan Smith.
 */

public class MacroParser {
    public static final char DEFINE_MACRO = '#';

    public static void addTo(ArrayList<Parser> parsers) {
        parsers.set(DEFINE_MACRO, MacroParser::parse);
    }

    public static NylonFunction parse(CharIterator in, NylonParser nylonParser) {
        in.skip();

        String name = ParseUtils.parseNextName(in);

        NylonFunction nylonFunction = nylonParser.parse(in);

        nylonParser.functions.put(name, () -> nylonFunction);

        return null;
    }
}

File: src/nylon/parser/parsers/MacroParser.kt
package nylon.parser.parsers

import nylon.NylonFunction
import nylon.parser.CharIterator
import nylon.parser.NylonParser
import nylon.parser.Parser
import nylon.parser.ParserBuilder
import java.util.*
import java.util.function.Supplier

/**
 * Created by Aedan Smith.
 */

class MacroParserBuilder : ParserBuilder {
    override fun accept(parsers: HashMap<Char, Parser>) {
        parsers['#'] = MacroParser()
    }
}

class MacroParser : Parser {
    override fun apply(src: CharIterator, parser: NylonParser): NylonFunction? {
        src.next()
        val name = src.parseNextName()
        val function = parser.parse(src)
        parser.functions.put(name, Supplier { function!!.clone() })
        println(name)
        return null
    }
}
